certbot/tests/main_test.py

```diff
@@ -539,7 +539,16 @@ def test_plugins_no_args(self, _det, mock_disco):
         ifaces = []
         plugins = mock_disco.PluginsRegistry.find_all()
 
-        _, stdout, _, _ = self._call(['plugins'])
+        stdout = six.StringIO()
```

Comment:
We ended up using code like this a lot more than I thought we would when I wrote [this comment](https://github.com/certbot/certbot/pull/5028#discussion_r134325650). Sorry for the mixed signals on this, but it'd be good if we factored it out to avoid duplication.

What if we made another function in `certbot/tests/util.py` called something like `patch_get_utility_and_stdout` than returned both the mock `getUtility` and a `StringIO` object that all `IDisplay` methods will write to? My understanding based on [this](http://www.voidspace.org.uk/python/mock/patch.html#where-to-patch) is to ensure writing messages works in all cases, we need to either have the function return the `StringIO` object or have the caller mock `certbot.tests.util.sys.stdout` which feels much more awkward.

Reply:
@bmw, Does this look acceptable?

```diff
diff --git a/certbot/tests/util.py b/certbot/tests/util.py
index 69896251..a61d6321 100644
--- a/certbot/tests/util.py
+++ b/certbot/tests/util.py
@@ -14,6 +14,7 @@ from cryptography.hazmat.backends import default_backend
 from cryptography.hazmat.primitives import serialization
 import mock
 import OpenSSL
+import six
 from six.moves import reload_module  # pylint: disable=import-error
 
 from acme import jose
@@ -169,6 +170,14 @@ def patch_get_utility(target='zope.component.getUtility'):
     return mock.patch(target, new_callable=_create_get_utility_mock)
 
 
+def patch_get_utility_with_stdout(target='zope.component.getUtility',
+                                  stdout=None):
+    stdout = stdout if stdout else six.StringIO()
+
+    callable_ = _create_get_utility_mock_with_stdout(stdout)
+    return mock.patch('zope.component.getUtility', new_callable=callable_)
+
+
 class FreezableMock(object):
     """Mock object with the ability to freeze attributes.
 
@@ -249,6 +258,24 @@ def _create_get_utility_mock():
     return FreezableMock(frozen=True, return_value=display)
 
 
+def _create_get_utility_mock_with_stdout(stdout):
+    def _write_msg(message, *args, **kwargs):
+        _, _ = args, kwargs
+        stdout.write(message)
+
+    display = FreezableMock()
+    for name in interfaces.IDisplay.names():  # pylint: disable=no-member
+        if name == 'notification':
+            setattr(display, name, _write_msg)
+        else:
+            frozen_mock = FreezableMock(frozen=True,
+                                        func=_assert_valid_call)
+            setattr(display, name, frozen_mock)
+    display.freeze()
+
+    return lambda: FreezableMock(frozen=True, return_value=display)
+
+

```

With the above changes, the `certbot.tests.main_test.MainTest.test_plugins_no_args` will look like:

```python
...
    @mock.patch('certbot.main.plugins_disco')
    @mock.patch('certbot.main.cli.HelpfulArgumentParser.determine_help_topics')
    def test_plugins_no_args(self, _det, mock_disco):
        ifaces = []
        plugins = mock_disco.PluginsRegistry.find_all()

        stdout = six.StringIO()
        with test_util.patch_get_utility_with_stdout(stdout=stdout) \
          as mock_get_utility:
            _, stdout, _, _ = self._call(['plugins'], stdout)

        plugins.visible.assert_called_once_with()
        plugins.visible().ifaces.assert_called_once_with(ifaces)
        filtered = plugins.visible().ifaces()
        self.assertEqual(stdout.getvalue().strip(), str(filtered))
...
```

```bash
(venv)  ~/v/git/others/certbot/certbot/tests 
↳ python main_test.py MainTest.test_plugins_no_args                                                                                                                                                       s@cygnus 
.                                                  
----------------------------------------------------------------------
Ran 1 test in 0.078s

OK

```